Ladybird/AppKit/UI/Inspector.mm

```diff
@@ -0,0 +1,266 @@
+
+            if (strong_self.dom_tree) {
+                [strong_self.dom_tree_outline_view reloadItem:nil reloadChildren:YES];
+                [strong_self.dom_tree_outline_view sizeToFit];
+            } else {
+                strong_self.dom_tree = @{};
+            }
+        };
+
+        [self setContentView:split_view];
+        [self setTitle:@"Inspector"];
+        [self setIsVisible:YES];
+
+        auto split_view_height = [split_view frame].size.height;
+        [split_view setPosition:(split_view_height * 0.6f) ofDividerAtIndex:0];
+    }
+
+    return self;
+}
+
+- (void)dealloc
+{
+    auto& web_view = [[self.tab web_view] view];
+    web_view.on_received_dom_tree = nullptr;
+}
+
+#pragma mark - Public methods
+
+- (void)inspect
+{
+    auto& web_view = [[self.tab web_view] view];
+    web_view.inspect_dom_tree();
+}
+
+- (void)reset
+{
+    self.dom_tree = @{};
+    [self.dom_tree_outline_view reloadItem:nil reloadChildren:YES];
+    [self.dom_tree_outline_view sizeToFit];
+}
+
+#pragma mark - Private methods
+
+- (void)initializeDOMTreeTab:(NSTabView*)tab_view
+{
+    auto* tab = [[NSTabViewItem alloc] initWithIdentifier:@"DOM Tree"];
+    [tab setLabel:@"DOM"];
+
+    auto* scroll_view = [[NSScrollView alloc] init];
+    [scroll_view setHasVerticalScroller:YES];
+    [scroll_view setHasHorizontalScroller:YES];
+    [scroll_view setLineScroll:24];
+    [tab setView:scroll_view];
+
+    self.dom_tree_outline_view = [[NSOutlineView alloc] initWithFrame:[tab_view frame]];
+    [self.dom_tree_outline_view setDoubleAction:@selector(onTreeDoubleClick:)];
+    [self.dom_tree_outline_view setDataSource:self];
+    [self.dom_tree_outline_view setDelegate:self];
+    [self.dom_tree_outline_view setHeaderView:nil];
+    [scroll_view setDocumentView:self.dom_tree_outline_view];
+
+    auto* column = [[NSTableColumn alloc] initWithIdentifier:@"DOM Tree"];
+    [self.dom_tree_outline_view addTableColumn:column];
+
+    [tab_view addTabViewItem:tab];
+}
+
+- (void)onTreeDoubleClick:(id)sender
+{
+    NSOutlineView* outline_view = sender;
+    id item = [outline_view itemAtRow:[outline_view clickedRow]];
+
+    if ([outline_view isItemExpanded:item]) {
+        [outline_view collapseItem:item];
+    } else {
+        [outline_view expandItem:item];
+    }
+}
+
+#pragma mark - NSOutlineViewDataSource
+
+- (id)outlineView:(NSOutlineView*)view
+            child:(NSInteger)index
+           ofItem:(id)item
+{
+    if (item == nil) {
+        item = self.dom_tree;
+    }
+
+    NSArray* children = [item objectForKey:@"children"];
+    if (children == nil)
+        return nil;
+
+    return [children objectAtIndex:index];
+}
+
+- (NSInteger)outlineView:(NSOutlineView*)view
+    numberOfChildrenOfItem:(id)item
+{
+    if (item == nil) {
+        item = self.dom_tree;
+    }
+
+    NSArray* children = [item objectForKey:@"children"];
+    if (children == nil)
+        return 0;
```

Comment:
You don't have to do these `nil` checks, this is Objective-C 😀

Reply:
Good point!